src/cosim/utility/thread_pool.hpp

```diff
@@ -0,0 +1,121 @@
+/**
+ *  \file
+ *  Slave interface.
+ *
+ *  \copyright
+ *      This Source Code Form is subject to the terms of the Mozilla Public
+ *      License, v. 2.0. If a copy of the MPL was not distributed with this
+ *      file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+#ifndef COSIM_UTILITY_THREAD_POOL_HPP
+#define COSIM_UTILITY_THREAD_POOL_HPP
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace cosim
+{
+namespace utility
+{
+
+class thread_pool
+{
+private:
+    bool done_;
+    std::queue<std::function<void()>> work_queue_;
+    std::vector<std::thread> threads_;
+    std::mutex m_;
+    std::condition_variable cv_finished_;
+    std::condition_variable cv_worker_;
+    unsigned int pending_tasks_;
+
+    void worker_thread()
+    {
+        while (true) {
+            std::unique_lock<std::mutex> lck(m_);
```

Comment:
Will it be possible to move this out of while loop so we don't create/destroy lck for each loop iteration?

Reply:
No, I don't think so. How do you then reacquire the lock after a task has been completed, if it's outside of this loop? You'd have to let the thread terminate and recreate it, which we don't want?